documentation/source/extensions.rst

```diff
@@ -0,0 +1,104 @@
+The module can then be installed with ``pip3 install cocotbext-spi``, and used with ``import cocotbext.spi``.
+
+Types of cocotb extensions
+--------------------------
+
+For some types of cocotb extensions we have developed conventions which go beyond naming.
+These conventions help to achieve a consistent behavior across extensions of the same type.
+
+Bus extensions
+~~~~~~~~~~~~~~
+
+A cocotb extension which interacts with a bus or an interface (such as SPI or AXI) should build on top of a common set of classes to provide a uniform interface for its users.
+Typically, a bus extension provides three pieces of functionality:
+an abstraction of the bus itself, a bus driver, and a bus monitor.
+
+A bus driver is the "active" part, it drives the signals that make up the bus to request reads or writes from the bus.
+A bus monitor is the "passive" part, it observes signal changes on the bus and assigns meaning to them.
+Monitors can also check the bus behavior against a standard to ensure no invalid states are being observed.
+
+The signals which make up the bus should be grouped in a class inheriting from :any:`cocotb.bus.Bus`.
+Bus drivers should inherit from the :any:`cocotb.drivers.BusDriver` class.
+Bus monitors should inherit from the :any:`cocotb.monitors.BusMonitor` class.
+
+Packaging extensions
+--------------------
+
+To package a cocotb extension as Python package follow the `Naming conventions`_, and the `normal Python packaging rules <https://packaging.python.org/tutorials/packaging-projects/>`_.
+
+The module file hierarchy should be as following (replace ``EXTNAME`` with the name of the extension, e.g. ``spi``).
+
+.. code-block::
+
+  ├── cocotbext/
+  │   └── EXTNAME/
+  │       └── __init__.py
+  ├── README.md
+  └── setup.py
+
+The Python source code should go into the ``EXTNAME`` directory, next to the ``__init__.py`` file.
+All packaging metadata goes into ``setup.py``.
+
+.. code-block:: python3
+
+  # setup.py
+  from setuptools import setup
+
+  setup(name='cocotbext-EXTNAME',
+        version='0.1',
+        install_requires = ['cocotb'],
+        classifiers = [
+          "Programming Language :: Python :: 3",
+          "License :: OSI Approved :: MIT License",
+          "Operating System :: OS Independent",
+          "Topic :: Scientific/Engineering :: Electronic Design Automation (EDA)"])
+
+With this file structure in place the cocotb extension can be installed through ``pip`` in development mode ::
+
+  $ python3 -m pip install -e .
+
+Once the extension has been `uploaded to PyPi <https://packaging.python.org/tutorials/packaging-projects/#uploading-the-distribution-archives>`_, it can be installed by name.
+
+.. code-block: command
+
+  $ python3 -m pip install cocotbext-EXTNAME
+
+To use the functionality in the extension module, import it into your testbench.
+
+.. code-block:: python3
+
+  import cocotbext.EXTNAME
```

Comment:
Perhaps worth mentioning also `from cocotbext import EXTNAME` and `from cocotb.EXTNAME import MyVerificationClass`